coordinator/src/substrate/db.rs

```diff
@@ -9,12 +9,38 @@ use serai_client::{
   validator_sets::primitives::{Session, ValidatorSet, KeyPair},
 };
 
-create_db! {
-  NewSubstrateDb {
+// This guarantees:
+// 1) Appended transactions are appended
+// 2) Taking cosigns does not clear any TXs which weren't taken
+// 3) Taking does actually clear the set
+static COSIGN_LOCK: OnceLock<Mutex<()>> = OnceLock::new();
+pub struct CosignTxn<T: DbTxn>(T, MutexGuard<'static, ()>);
+impl<T: DbTxn> CosignTxn<T> {
+  pub fn new(txn: T) -> Self {
+    Self(txn, COSIGN_LOCK.get_or_init(|| Mutex::new(())).lock().unwrap())
+  }
+  pub fn commit(self) {
+    self.0.commit();
+  }
+}
+
+create_db!(
+  CoordinatorSubstrateDb {
     CosignTriggered: () -> (),
     IntendedCosign: () -> (u64, Option<u64>),
     BlockHasEvents: (block: u64) -> u8,
-    CosignTransactions: (network: NetworkId) -> Vec<(Session, u64, [u8; 32])>
+    LatestCosignedBlock: () -> u64,
+    CosignTransactions: (network: NetworkId) -> Vec<(Session, u64, [u8; 32])>,
+    BlockDb: () -> u64,
```

Comment:
Mind if we call this NextBlock?

Reply:
Looks good to me, I’m not that opinionated 😄